File: wqi.js
const axios = require('axios');
const opencage = require('opencage-api-client');

const opencageApiKey = 'cc5196c709b9434a9c993fec31dd4820'; // Replace with your OpenCage API key
const meersensApiKey = '4FYIeuwcXmpspTIjK4fkK8BJJIHSpwuL'; // Replace with your Meersens API key

function geocodeAddress(address) {
  return new Promise((resolve, reject) => {
    opencage
      .geocode({ q: address, key: opencageApiKey })
      .then((data) => {
        if (data.status.code === 200 && data.results.length > 0) {
          const place = data.results[0];
          const latitude = place.geometry.lat;
          const longitude = place.geometry.lng;
          resolve({ latitude, longitude });
        } else {
          reject('Invalid address or no results found');
        }
      })
      .catch((error) => {
        reject('Error while geocoding');
      });
  });
}

const wqiData = async (city) => {
  let lat, lng;

  try {
    const coords = await geocodeAddress(city);
    lat = coords.latitude;
    lng = coords.longitude;
  } catch (error) {
    console.error('Error geocoding address:', error);
    return;
  }

  if (lat !== undefined && lng !== undefined) {
    const url = `https://api.meersens.com/environment/public/air/current?lat=${lat}&lng=${lng}`;

    try {
      const response = await axios.get(url, {
        headers: {
          'apikey': meersensApiKey,
        },
      });

      if (response.status === 200) {
        const data = response.data;
        if (data && data.found) {
          console.log('\nIndex type:', data.index.index_type);
          console.log('Overall Qualification:', data.index.qualification);
          console.log('Total index value:', data.index.value);
          console.log('Main pollutants list:');
          console.log(data.index.main_pollutants.join(', '));
          console.log('#'.repeat(20));

          const pollutants = data.pollutants;
          for (const key in pollutants) {
            if (pollutants.hasOwnProperty(key)) {
              const pollutant = pollutants[key];
              console.log(`Chemical name: ${pollutant.name} (${pollutant.shortcode})`);
              console.log(`Chemical index value: ${pollutant.value}`);
              console.log(`Confidence level: ${pollutant.confidence}`);
              console.log(`Index Type: ${pollutant.index.index_type}`);
              console.log(`Index Name: ${pollutant.index.index_name}`);
              console.log(`Qualification: ${pollutant.index.qualification}`);
              console.log(`Description: ${pollutant.index.description}`);
              console.log(`Overall chemical index level: ${pollutant.index.value}`);
              console.log('*'.repeat(20));
            }
          }
        } else {
          console.log(data);
        }
      } else {
        console.error(`Failed to fetch WQI data for city: ${city}`);
      }
    } catch (error) {
      console.error('Error fetching WQI data:', error.message);
    }
  }
};

const city = 'Munich'; // Replace with your desired city

wqiData(city);

File: wqi.py
import requests
from geopy.geocoders import Nominatim
# it get latitude and longitude value by area name 
def get_lat_lng(city: str) -> tuple:
   geolocator = Nominatim(user_agent="your_app_name")
   location = geolocator.geocode(city)
   if location:
      latitude = round(location.latitude, 3)
      longitude = round(location.longitude, 3)
      return latitude, longitude
   return 0, 0

# it gets water quality index from api using api_key and return that data 
def wqi_data(city: str) -> dict:
   api_key = "4FYIeuwcXmpspTIjK4fkK8BJJIHSpwuL"
   lat, lng = get_lat_lng(city)
   if lat and lng:
      url = f"https://api.meersens.com/environment/public/air/current?lat={lat}&lng={lng}"
      headers = {
            "apikey": api_key
      }
      response = requests.get(url, headers=headers)
      if response.status_code == 200:
         data = response.json()
         return data
   return {"found": False}

# Enter the city name 
city = input("city name: ").lower().strip()
# get the  wqi details using api details

details = wqi_data(city)

# parse the data and print it on screen you can manuplate it accoding to what do you want to do with this data.
if details['found']:
   print("\n")
   print(f"Date time: {details['datetime']}")
   print(f"Index type: {details['index']['index_type']}")
   print(f"Overall Qualificaton: {details['index']['qualification']}")
   print(f"Total index value: {details['index']['value']}")
   print(f"Main pollutaints list: ")
   print(*details['index']['main_pollutants'])
   
   print(f"{'#'*20}\n")

   pollutants_keys = details.get('pollutants', {})
   for pollutant_key, pollutant_data in pollutants_keys.items():
      print(f"Chemical name: {pollutant_data['name']} ({pollutant_data['shortcode']})")
      print(f"Chemical index value: {pollutant_data['value']}")
      print(f"Confidence level: {pollutant_data['confidence']}")
      print(f"Index Type: {pollutant_data['index']['index_type']}")
      print(f"Index Name: {pollutant_data['index']['index_name']}")
      print(f"Qualification: {pollutant_data['index']['qualification']}")
      print(f"Description: {pollutant_data['index']['description']}")
      print(f"Overall chemical index level: {pollutant_data['index']['value']}")
      print(f"{'*'*20}\n")
else:
   print(details)
